packages/flutter/lib/src/material/material_localizations.dart

```diff
@@ -99,25 +111,30 @@ abstract class MaterialLocalizations {
 
 /// Localized strings for the material widgets.
 class DefaultMaterialLocalizations implements MaterialLocalizations {
-  /// Construct an object that defines the material widgets' localized strings
+  /// Constructs an object that defines the material widgets' localized strings
   /// for the given `locale`.
   ///
   /// [LocalizationsDelegate] implementations typically call the static [load]
   /// function, rather than constructing this class directly.
-  DefaultMaterialLocalizations(this.locale) {
+  factory DefaultMaterialLocalizations(Locale locale) {
     assert(locale != null);
-    _nameToValue = localizations[_localeName]
-      ?? localizations[locale.languageCode]
-      ?? localizations['en']
-      ?? <String, String>{};
+
+    final Map<String, String> result = <String, String>{};
+    if (localizations.containsKey(locale.languageCode))
+      result.addAll(localizations[locale.languageCode]);
+    if (localizations.containsKey(locale.toString()))
+      result.addAll(localizations[locale.toString()]);
+    return new DefaultMaterialLocalizations._(locale, result);
```

Comment:
We need to document this convention.

Now that I better understand what this is doing, I think it would be better to premultiply this fallback behaviour in the generator script. It seems really weird to be doing it at runtime. It also seems to interfere with the supportedLocales logic, which tries to do its own locale fallback. Maybe @HansMuller has opinions on this.

Reply:
To avoid scope creep in this PR, I filed a separate issue for this: https://github.com/flutter/flutter/issues/12090.

As for dartdocs, I don't know how to refer to `.arb` files. It's all private implementation details.